### PR TITLE
[TEST] make AbstractQueryTestCase#testUnknownObjectException more accurate

### PR DESCRIPTION
**Pull request by javanna**
`testUnknownObjectException` used to generate malformed json objects in some cases, due to the existence of arrays as it was not closing the injected object correctly. That is why the test was catching `JsonParseException` among the exception expected to be thrown. That is fixed by tracking where the new object is placed and placing its end object marker to the right level rather than always at the end.

Also introduced a mechanism to explicitly declare objects that won't cause any exception when they get additional objects injected, so that there is no need to override the method anymore as that caused copy pasting of the whole test method. This also makes sure that changes are reflected in tests, as those inner objects are not skipped but we actually check that what is declared is true (no exceptions get thrown when an additional object is added within them.

### CODE REVIEW

**Comment by javanna** [comment]
@cbuescher can you have a look please?

**Comment by cbuescher** [comment]
@javanna I took a look, most of the change looks good to me. A suggestion though, because I found the insertion logic in `testUnknownObjectException()` quiet hard to grasp (even though I'm guilty of creating that test, sorry). I think it might be possible to do something similar than what we so when suffling the fields in a json object in ESTestCase#shuffleXContent(): we could take the valid query, parse it to an ordered map, modify that map so it contains the offending field that should cause the exception and then render that back to the query builder. Might be easier to maintain on the long run. I haven't looked into the details yet, if you want to keep inserting by traversing the query string I'm also fine with that and will take another look.

**Comment by javanna** [comment]
@cbuescher the previous version had numerous bugs :) I pushed an update. I agree that we should move away from concatenating strings in this text. I quickly gave it a try but it's quite a job. I left a TODO and we can always do it as a follow-up.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 1]
```diff
@@ -312,30 +312,115 @@ public void testUnknownField() throws IOException {
...
-    public void testUnknownObjectException() throws IOException {
+    public final void testUnknownObjectException() throws IOException {
         String validQuery = createTestQueryBuilder().toString();
+        unknownObjectExceptionTest(validQuery);
+        for (String query : getAlternateVersions().keySet()) {
+            unknownObjectExceptionTest(query);
+        }
+    }
+
+    private void unknownObjectExceptionTest(String validQuery) throws IOException {
```
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: Can we add the general idea about how this test is supposed to modify the original query here, maybe with a small example? e.g. we just talked about that the outer loop that traverses the whole json query  adds a new object within each object (after encountering an opening bracket, under certain conditions) of the whole big query. This doesn't need to account for all kinds of exceptions (like we are inside quotes etc...)

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 2]
```diff
@@ -312,30 +312,115 @@ public void testUnknownField() throws IOException {
...
+        //TODO building json by concatenating strings makes the code unmaintainable, we should rewrite this test
         assertThat(validQuery, containsString("{"));
+        int level = 0;
+        boolean withinQuotes = false;
```
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: maybe a small comment why we keep track of whether we are inside quotes

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 3]
```diff
@@ -312,30 +312,115 @@ public void testUnknownField() throws IOException {
...
+                if (objectHoldingArbitraryContentLevel == 0) {
+                    //reset the flag once we have traversed the whole object that holds arbitrary content
+                    expectedException = true;
+                }
+            } else if (withinQuotes == false && validQuery.charAt(insertionPosition) == '{') {
+                level++;
+                if (expectedException) {
+                    int start = -1;
+                    int end = -1;
+                    for (int i = insertionPosition; i >= 0; i--) {
```
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: can we add a small comment that this tries to detect the fieldname before the current opening bracket, to check if we are in a block that can contain arbitrary json content? At least thats what I think this is doing...

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 4]
```diff
@@ -312,30 +312,115 @@ public void testUnknownField() throws IOException {
...
+                    }
+                    if (start >= 0  && end > 0) {
+                        String objectName = validQuery.substring(start, end);
+                        expectedException = getObjectsHoldingArbitraryContent().contains(objectName) == false;
+                    }
+                }
+                if (expectedException == false) {
+                    objectHoldingArbitraryContentLevel++;
+                }
+                String testQuery = validQuery.substring(0, insertionPosition) + "{ \"newField\" : ";
```
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: Maybe point out that this is actually the place where we modify the valid input query by adding a new object level to it.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java [hunk 5]
```diff
@@ -312,30 +312,115 @@ public void testUnknownField() throws IOException {
...
+                String secondPart = validQuery.substring(insertionPosition);
+                int currentLevel = level;
+                boolean quotes = false;
+                for (int i = 0; i < secondPart.length(); i++) {
```
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: This loop tries to insert the closing bracket at the same level as the newly inserted object, right? Also maybe add a short reminder about that in front of the loop so its a bit quicker to figure that out again later.

**Comment by cbuescher** [comment]
@javanna agree that this test it somewhat complex in its current form and agree we should keep a reminder to simplify as a follow up. In the meantime I made some suggestions about adding comments in places that I found a bit hard to follow when I tried reading the test method. This doesn't have to be exhaustive comments but maybe add some help to structure the way the test works when reading this again as our future selves.

**Comment by cbuescher** [comment]
@javanna thanks for adding those great comments

**Comment by tlrx** [comment]
> I agree that we should move away from concatenating strings in this test. I quickly gave it a try but it's quite a job. I left a TODO and we can always do it as a follow-up.

CI found a bug with this method so I created https://github.com/elastic/elasticsearch/pull/20100 to fix it and change the whole method so that it is no more based on string concatenation.
